src/zircon/chunkserver/control/handle.go

```diff
@@ -3,10 +3,331 @@ package control
+		return fmt.Errorf("attempt to create duplicate chunk: %d/%d", chunk, initialVersion)
+	}
+	err = cs.Storage.WriteVersion(chunk, initialVersion, initialData)
+	if err != nil {
+		return err
+	}
+	err = cs.Storage.SetLatestVersion(chunk, initialVersion)
+	if err != nil {
+		err2 := cs.Storage.DeleteVersion(chunk, initialVersion)
+		if err2 != nil {
+			panic("failed to be able to maintain invariant") // TODO: handle this more gracefully than crashing
+		}
+		return err
+	}
+	return nil
+}
+
+func (cs *chunkserver) Delete(chunk apis.ChunkNum, version apis.Version) error {
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
+
+	if version <= 0 {
+		return fmt.Errorf("deleted version was not positive: %d/%d", chunk, version)
+	}
+
+	latest, err := cs.Storage.GetLatestVersion(chunk)
+	if err != nil {
+		return err
+	}
+
+	// if we delete the latest version, we also delete everything newer... and because nothing older will exist at this
+	// point, we delete everything.
+	if latest == version {
+		versions, err := cs.Storage.ListVersions(chunk)
+		if err != nil {
+			return err
+		}
+		// mark the entire chunk as able to be deleted
+		if err := cs.Storage.DeleteLatestVersion(chunk); err != nil {
+			return err
+		}
+		// then delete all versions of the chunk
+		for _, delver := range versions {
+			if err := cs.Storage.DeleteVersion(chunk, delver); err != nil {
+				return err
+			}
+		}
+	} else {
+		// just delete the single version
+		if err := cs.Storage.DeleteVersion(chunk, version); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (cs *chunkserver) Teardown() {
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
+
+	// wipe away any pending hashes
+	// TODO: have a way to regularly wipe away stale pending hashes
+	cs.Hashes = map[apis.CommitHash]commit{}
+}
+
+// Given a chunk reference, read out part or all of a chunk.
+// If 'minimum' is AnyVersion, then whichever version the chunkserver currently has will be returned.
+// If the version of the chunk that this chunkserver has is at least the minimum version, it will be returned.
+// Otherwise, an error will be returned, along with the most recent available version.
+// The sum of offset + length must not be greater than MaxChunkSize. The number of bytes returned is always exactly
+// the same number of bytes requested if there is no error.
+// The version of the data actually read will be returned.
+// Fails if a copy of this chunk isn't located on this chunkserver.
+func (cs *chunkserver) Read(chunk apis.ChunkNum, offset apis.Offset, length apis.Length, minimum apis.Version) ([]byte, apis.Version, error) {
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
+
+	if apis.Length(offset) + length > apis.MaxChunkSize {
+		return nil, 0, errors.New("too much data")
+	}
+
+	version, err := cs.Storage.GetLatestVersion(chunk)
+	if err != nil {
+		return nil, 0, err
+	}
+	if version < minimum {
+		return nil, version, errors.New("requested newer version than was available")
+	}
+	data, err := cs.Storage.ReadVersion(chunk, version)
```

Comment:
Does this mean we read the whole block every time we want to read part of a block?

Reply:
The idea is that the disk layer includes caching.